state/txindex/kv/kv.go

```diff
@@ -46,13 +54,322 @@ func (txi *TxIndex) Get(hash []byte) (*types.TxResult, error) {
 	return txResult, nil
 }
 
-// AddBatch writes a batch of transactions into the TxIndex storage.
-func (txi *TxIndex) AddBatch(b *txindex.Batch) error {
+// AddBatch indexes a batch of transactions using the given list of tags.
+func (txi *TxIndex) AddBatch(b *txindex.Batch, allowedTags []string) error {
 	storeBatch := txi.store.NewBatch()
+
 	for _, result := range b.Ops {
-		rawBytes := wire.BinaryBytes(&result)
-		storeBatch.Set(result.Tx.Hash(), rawBytes)
+		hash := result.Tx.Hash()
+
+		// index tx by tags
+		for _, tag := range result.Result.Tags {
+			if stringInSlice(tag.Key, allowedTags) {
+				storeBatch.Set(keyForTag(tag, result), hash)
+			}
+		}
+
+		// index tx by hash
+		rawBytes := wire.BinaryBytes(result)
+		storeBatch.Set(hash, rawBytes)
 	}
+
 	storeBatch.Write()
 	return nil
 }
+
+// Index indexes a single transaction using the given list of tags.
+func (txi *TxIndex) Index(result *types.TxResult, allowedTags []string) error {
+	batch := txindex.NewBatch(1)
+	batch.Add(result)
+	return txi.AddBatch(batch, allowedTags)
+}
+
+func (txi *TxIndex) Search(q *query.Query) ([]*types.TxResult, error) {
+	hashes := make(map[string][]byte) // key - (base 16, upper-case hash)
+
+	// get a list of conditions (like "tx.height > 5")
+	conditions := q.Conditions()
+
+	// if there is a hash condition, return the result immediately
+	hash, err, ok := lookForHash(conditions)
+	if err != nil {
+		return []*types.TxResult{}, errors.Wrap(err, "error during searching for a hash in the query")
+	} else if ok {
+		res, err := txi.Get(hash)
+		return []*types.TxResult{res}, errors.Wrap(err, "error while retrieving the result")
+	}
+
+	// conditions to skip
+	skipIndexes := make([]int, 0)
+
+	// if there is a height condition ("tx.height=3"), extract it for faster lookups
+	height, heightIndex := lookForHeight(conditions)
+	if heightIndex >= 0 {
+		skipIndexes = append(skipIndexes, heightIndex)
+	}
+
+	var hashes2 [][]byte
+
+	// extract ranges
+	// if both upper and lower bounds exist, it's better to get them in order not
+	// no iterate over kvs that are not within range.
+	ranges, rangeIndexes := lookForRanges(conditions)
+	if len(ranges) > 0 {
+		skipIndexes = append(skipIndexes, rangeIndexes...)
+	}
+	for _, r := range ranges {
+		hashes2 = txi.matchRange(r, startKeyForRange(r, height, heightIndex > 0))
```

Comment:
do we need to pass `heightSpecified` ? Isn't checking if the `height == 0` enough ?

Reply:
yeah, but height == 0 looks kinda fishy...